subcommands/targets/offline-update.go

```diff
@@ -366,6 +375,44 @@ func downloadTufRepo(factory string, target string, tag string, prod bool, wave
 	return nil
 }
 
+func copyRecursive(src, dst string) error {
+	copyFile := func(src, dst string) error {
+		srcFile, err := os.Open(src)
+		if err != nil {
+			return nil
+		}
+		defer srcFile.Close()
+
+		dstFile, err := os.Create(dst)
```

Comment:
(2) probably makes sense... although I'm not a great fan of a trade-off (added complexity).

As for (1), I think it comes with inherent risk of copying wrong files.
Think of it this way: a user might supply a wrong source, interrupt, and then supply a correct source.
If we don't forcibly replace all files, we can end up with the wrong files.

Not sure how accurate are my statements. Maybe, ostree has a directory structure preventing the above.
So, I leave it up to you to weigh on this.

Reply:
(2) is important if a user copies files to external thumb drive to guarantee data are synced.

Reply:
>a user might supply a wrong source, interrupt, and then supply a correct source.

I don't see how (1) would cause an issue here taking into account (2) and the fact that each file in ostree repo has unique name - its hash.

Reply:
> each file in ostree repo has unique name - its hash

Told you, I can be missing something.

> how (1) would cause an issue here

But, I must be backing my claims about safety still.
You see, fsync only makes sure to sync the write to the hardware level, but that data may still be stored in the hardware (firmware) level cache first, and only later written to the persistent storage.
So, if we are talking about e.g. a thumb drive, than I'd say to do one of the following:
1. Always rewrite - easy way.
2. If file exists - check the file size/hash, and rewrite if it differs.

Trust me, I remember those times when I had to trash 1 in 100 CDs because the fsync is not as reliable as assumed.

Reply:
check&write vs rewrite and "write to tmp" and fsync are two different things.
If we write to `tmp` then a file does not exist until it's fsynced and renamed to the file correct name. If we assume that "fsync and rename" are 99.9999% guarantee that data are persisted then we don't need to check file hashes. We use this techniques in fioconfig and aklite and composectl and we haven't seen any issues with it so far.

Reply:
Yes, you're right.
Fsync to tmp and then rename should do the trick.

Probably in this specific use case (writing a lot of small files) it is worth an associated complexity.

Reply:
Now using .tmp file + sync + move.